### DISCUSSION TITLE
Low FPS despite low GPU usage

### DISCUSSION THREAD

**Discussion by Cyliann** [bevyengine/bevy#6112] [Q&A] [answered]
I know this is very similar to [#1410](https://github.com/bevyengine/bevy/discussions/1410#discussion-2982121), but that one was project specific and the solution didn't work for me.

I'm trying to make a Minecraft like [voxel engine](https://github.com/Cyliann/minecrust) and just finished basic terrain generation. The problem is with as little as 64x64x16 blocks the performance is terrible. I'm generating my own meshes (it's probably cheaper than Cube shape) and do occlusion culling (removing hidden faces). Even when compiling with `--release` flag I get <10fps. GPU and CPU usage is around 10-20%. 

Is there anything I can do to improve the performance?

**Comment by bjorn3** [comment]
Have you tried profiling it using https://github.com/wolfpld/tracy? That may help finding the bottleneck. I first thought that PCIe bandwidth could be the issue, but a quick calculation[^1] shows that only ~750MB/s is used at 10fps assuming re-upload of everything on every frame. I believe Bevy doesn't re-upload if meshes aren't changed, so the PCIe bandwidth should be even lower.

[^1]: 64\*64\*16 (voxel count) * 6\*2 (triangles per voxel) * 3 (vertexes per triangle) * 4\*4 (bytes per vec4) * 2 (position+normal) * 10 (frames per second)

**Reply by Cyliann** [reply]
> Have you tried profiling it using https://github.com/wolfpld/tracy? That may help finding the bottleneck.
 
I'll take a look at it and come back with results. Thanks

**Comment by Cyliann** [comment]
Important note: When I'm looking away from the mesh I easily get 60 fps, so it's a matter of rendering.

**Comment by Cyliann** [comment, marked as answer]
I didn't manage to get `trace_tracy` working due to `The client you are trying to connect to uses incompatible protocol version.`, but using `trace_chrome` I got [this](http://0x0.st/o4sI.json).

Also using cargo's `flamegraph` I managed to get this:
![flamegraph](https://user-images.githubusercontent.com/33868960/192581629-2d7d170c-985d-41a2-bc16-1a24c03234f8.svg)

**Reply by bjorn3** [reply]
https://github.com/nagisa/rust_tracy_client has a compatibility table between tracing_tracy crate versions and versions of the tracy program.

**Reply by Cyliann** [reply]
Ok. I finally did it. It shows exactly the same results as trace_chrome. `main_opaque_pass_3d` takes A LOT of time (great majority). Any ideas how to eliminate it?
![image](https://user-images.githubusercontent.com/33868960/192620518-c26b068a-ad59-4b65-bfbd-1c3c6755f798.png)

**Reply by bjorn3** [reply]
@superdump?

**Reply by superdump** [reply]
That seems odd. How exactly are you drawing stuff? As in, is that from drawing one single mesh? If so, that should mean creating one single draw command which should be very fast.

Ok, I checked the linked code and in create_mesh which is called from spawn_chunk you spawn a single Mesh per plane face of the cube for 64^3 cubes. That’s a lot. It’s important to try to minimise the number of entities with Meshes and so draw commands for performance. Making a single Mesh per chunk will help a lot.

**Reply by Cyliann** [reply]
After a little rework I managed to do single mesh per chunk. Now it works at stable 60fps. Thanks a lot
